### PR TITLE
Fix Eth1 data underflow

### LINKED ISSUE #540: Eth1 data underflow
https://github.com/sigp/lighthouse/blob/192380cb58c4e0803f9b787d033bc78fd0e9ea49/eth2/state_processing/src/per_block_processing.rs#L380-L383

If we accept that `state.eth1_data` is user-supplied input, I believe this subtraction is an underflow risk.

@michaelsproul, I'm keen to hear your thoughts.

### PR DESCRIPTION
**Pull request by ackintosh**
## Issue Addressed

This PR fixes #540. :octocat:

### CODE REVIEW

**Comment by ackintosh** [comment]
I'll add a test for the method added in this PR. 💨

File: eth2/types/src/beacon_state.rs [hunk 1]
```diff
@@ -764,6 +768,20 @@ impl<T: EthSpec> BeaconState<T> {
         })
     }
 
+    /// Get the number of outstanding deposits.
+    ///
+    /// Returns `Err` if the state is invalid.
+    pub fn get_outstanding_deposit_len(&self) -> Result<u64, Error> {
+        if let Some(len) = self.eth1_data.deposit_count.checked_sub(self.eth1_deposit_index) {
```
Comment (paulhauner) [inline eth2/types/src/beacon_state.rs:775]: Perhaps this is a good fit for (excuse lazy indenting/formatting):

```rust
self.eth1_data
    .deposit_count
    .checked_sub(self.eth1_deposit_index)
    .ok_or_else(|| {
          Err(Error::InvalidDepositState {
              deposit_count: self.eth1_data.deposit_count,
              deposit_index: self.eth1_deposit_index,
          })
    })
```

What do you think? :)

Comment (ackintosh) [reply eth2/types/src/beacon_state.rs:775]: It's cool, thank you!

**Comment by ackintosh** [comment]
This PR is ready for review. 😃

**Pull request merged**